lib/Parser/parse.cpp

```diff
@@ -4449,17 +4432,28 @@ bool Parser::ParseFncDeclHelper(ParseNodePtr pnodeFnc, ParseNodePtr pnodeFncPare
                 pnodeFnc->sxFnc.pnodeVars = nullptr;
                 m_ppnodeVar = &pnodeFnc->sxFnc.pnodeVars;
 
-                // We can't merge the param scope and body scope any more as the nested methods may be capturing params.
-                Scope* paramScope = pnodeFnc->sxFnc.pnodeScopes->sxBlock.scope;
-                if (!paramScope->GetCanMergeWithBodyScope())
+                if (!pnodeFnc->sxFnc.IsSimpleParameterList() && !fAsync)
                 {
-                    Assert(!pnodeFnc->sxFnc.IsSimpleParameterList());
-                    OUTPUT_TRACE_DEBUGONLY(Js::ParsePhase, L"The param and body scope of the function %s cannot be merged\n", pnodeFnc->sxFnc.pnodeName ? pnodeFnc->sxFnc.pnodeName->sxVar.pid->Psz() : L"Anonymous function");
-                    // Now add a new symbol reference for each formal in the param scope to the body scope.
-                    paramScope->ForEachSymbol([this](Symbol* param) {
-                        OUTPUT_TRACE_DEBUGONLY(Js::ParsePhase, L"Creating a duplicate symbol for the parameter %s in the body scope\n", param->GetPid()->Psz());
-                        this->CreateVarDeclNode(param->GetPid(), param->GetSymbolType(), false, nullptr, false);
+                    // We can't merge the param scope and body scope if the nested methods within the param scope captures any param.
+                    Scope* paramScope = pnodeFnc->sxFnc.pnodeScopes->sxBlock.scope;
+                    paramScope->ForEachSymbol([paramScope](Symbol* sym) {
+                        if (sym->GetPid()->GetTopRef()->sym == nullptr && paramScope->GetCanMergeWithBodyScope())
```

Comment:
Finding the sym has nonlocal reference based on topref does not have sym seemed too funky to me. Isn't this fragile?
I'll leave this to @pleath if he is okay with it.